executor/simple.go

```diff
@@ -976,6 +1085,256 @@ func (e *SimpleExec) executeCreateUser(ctx context.Context, s *ast.CreateUserStm
 	return domain.GetDomain(e.ctx).NotifyUpdatePrivilege()
 }
 
+func getUserPasswordLimit(ctx context.Context, sqlExecutor sqlexec.SQLExecutor, name string, host string, passwdlockinfo *passwordLockInfo) (*passwordReuseInfo, error) {
+	res := &passwordReuseInfo{notSpecified, notSpecified}
+	sql := new(strings.Builder)
+	sqlexec.MustFormatSQL(sql, `SELECT Password_reuse_history,Password_reuse_time FROM %n.%n WHERE User=%? AND Host=%?;`,
+		mysql.SystemDB, mysql.UserTable, name, strings.ToLower(host))
+	// Query the specified user password reuse rules
+	recordSet, err := sqlExecutor.ExecuteInternal(ctx, sql.String())
+	if err != nil {
+		return nil, err
+	}
+	req := recordSet.NewChunk(nil)
+	err = recordSet.Next(ctx, req)
+	var rowsNum int = 0
+	if err == nil {
+		rowsNum = req.NumRows()
+	}
+	errClose := recordSet.Close()
+	if errClose != nil {
+		return nil, errClose
+	}
+	if rowsNum != 1 {
+		err := fmt.Errorf("Unable to confirm `%s`@`%s` password reuse configuration information", name, strings.ToLower(host))
+		return nil, err
+	}
+	iter := chunk.NewIterator4Chunk(req)
+	for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+		if !row.IsNull(0) {
+			res.passwordHistory = int64(row.GetUint64(0))
+		} else {
+			res.passwordHistory = variable.PasswordHistory.Load()
+		}
+		if !row.IsNull(1) {
+			res.passwordReuseInterval = int64(row.GetUint64(1))
+		} else {
+			res.passwordReuseInterval = variable.PasswordReuseInterval.Load()
+		}
+	}
+	if passwdlockinfo.passwordHistoryFlag {
+		res.passwordHistory = passwdlockinfo.passwordHistory
+	}
+	if passwdlockinfo.passwordReuseIntervalFlag {
+		res.passwordReuseInterval = passwdlockinfo.passwordReuseInterval
+	}
+	return res, nil
+}
+
+// get the boundary of password valid time
+func getValidTime(sctx sessionctx.Context, passwordReuse *passwordReuseInfo) string {
+	nowTime := time.Now().In(sctx.GetSessionVars().TimeZone)
+	nowTimeS := nowTime.Unix()
+	beforeTimeS := nowTimeS - passwordReuse.passwordReuseInterval*24*int64(time.Hour/time.Second)
+	if beforeTimeS < 0 {
+		beforeTimeS = 0
+	}
+	return time.Unix(beforeTimeS, 0).Format("2006-01-02 15:04:05.999999999")
+}
+
+// deleteHistoricalData delete useless password history
+// The deleted password must meet the following conditions at the same time
+// 1. Exceeded the maximum number of saves
+// 2. The password has exceeded the prohibition time
+func deleteHistoricalData(ctx context.Context, sqlExecutor sqlexec.SQLExecutor, userDetail *userInfo, maxDelRows int64, passwordReuse *passwordReuseInfo, sctx sessionctx.Context) error {
+	//never times out or no row need delete
+	if (passwordReuse.passwordReuseInterval > math.MaxInt32) || maxDelRows == 0 {
+		return nil
+	}
+	sql := new(strings.Builder)
+	// no prohibition time
+	if passwordReuse.passwordReuseInterval == 0 {
+		deleteTemplate := `DELETE from %n.%n WHERE User= %? AND Host= %?  order by Password_timestamp ASC LIMIT `
```

Comment:
```suggestion
		deleteTemplate := `DELETE from %n.%n WHERE User= %? AND Host= %?  order by Password_timestamp ASC LIMIT `
```
Please remove this blank line.